### PR TITLE
add FalseClass and Time to the SafeMarshal list

### LINKED ISSUE #6614: bundler reports "could not fetch from the dependency API, trying the full index" when using packagecloud
### Describe the problem as clearly as you can

Bundler reports warning "Retrying dependency api due to error (x/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed." / "could not fetch from the dependency API, trying the full index" when downloading packages from packagecloud.io.

I haven't noticed it with bundler 2.2.x / ruby 2.7.x.
I'm wondering when this changed and if it's expected that packagecloud has to support an "/versions" endpoint.

```
bash-4.4# bundle install --jobs=8 --verbose
Running `bundle install --jobs 8 --verbose` with bundler 2.4.12

Found changes from the lockfile, re-resolving dependencies because the dependencies in your gemfile changed, you added a new platform to your gemfile
HTTP GET https://packagecloud.io/cinc-project/stable/versions
HTTP 404 Not Found https://packagecloud.io/cinc-project/stable/versions

Bundler::Fetcher::FallbackError: Net::HTTPNotFound: https://packagecloud.io/cinc-project/stable/versions
/opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher/downloader.rb:45:in `fetch'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher/compact_index.rb:124:in `call'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/compact_index_client/updater.rb:47:in `update'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/compact_index_client.rb:92:in `update'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/compact_index_client.rb:79:in `update_and_parse_checksums!'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher/compact_index.rb:68:in `available?'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher/compact_index.rb:15:in `call'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher/compact_index.rb:15:in `block in compact_index_request'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/fetcher.rb:158:in `use_api'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/source/rubygems.rb:392:in `block in api_fetchers'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/source/rubygems.rb:392:in `select'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/source/rubygems.rb:392:in `api_fetchers'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/source/rubygems.rb:293:in `dependency_api_available?'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:560:in `all?'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:560:in `precompute_source_requirements_for_indirect_dependencies?'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:854:in `source_requirements'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:497:in `resolution_packages'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:486:in `resolver'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:554:in `start_resolution'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:289:in `resolve'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:507:in `materialize'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/definition.rb:197:in `specs'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/installer.rb:231:in `ensure_specs_are_compatible!'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/installer.rb:83:in `block in run'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/process_lock.rb:12:in `block in lock'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/process_lock.rb:9:in `open'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/process_lock.rb:9:in `lock'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/installer.rb:71:in `run'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/installer.rb:23:in `install'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/cli/install.rb:62:in `run'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/cli.rb:261:in `block in install'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/settings.rb:131:in `temporary'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/cli.rb:260:in `install'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/cli.rb:34:in `dispatch'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/cli.rb:28:in `start'
  /opt/bundle/gems/bundler-2.4.12/exe/bundle:45:in `block in <top (required)>'
  /opt/bundle/gems/bundler-2.4.12/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
  /opt/bundle/gems/bundler-2.4.12/exe/bundle:33:in `<top (required)>'
  /opt/rbenv/versions/3.2.2/bin/bundle:25:in `load'
  /opt/rbenv/versions/3.2.2/bin/bundle:25:in `<main>'
HTTP GET https://packagecloud.io/cinc-project/stable/api/v1/dependencies
HTTP 200 OK https://packagecloud.io/cinc-project/stable/api/v1/dependencies
Fetching gem metadata from https://packagecloud.io/cinc-project/stable/
Query List: ["cinc-auditor-bin"]
Query Gemcutter Dependency Endpoint API: cinc-auditor-bin
HTTP GET https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
HTTP 200 OK https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
Retrying dependency api due to error (2/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.
Query Gemcutter Dependency Endpoint API: cinc-auditor-bin
HTTP GET https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
HTTP 200 OK https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
Retrying dependency api due to error (3/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.
Query Gemcutter Dependency Endpoint API: cinc-auditor-bin
HTTP GET https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
HTTP 200 OK https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
Retrying dependency api due to error (4/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.
Query Gemcutter Dependency Endpoint API: cinc-auditor-bin
HTTP GET https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
HTTP 200 OK https://packagecloud.io/cinc-project/stable/api/v1/dependencies?gems=cinc-auditor-bin
could not fetch from the dependency API, trying the full index

```

### Did you try upgrading rubygems & bundler?

I've upgraded to latest.

### Post steps to reproduce the problem

```
bash-4.4# ruby --version
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]
bash-4.4#
bash-4.4#
bash-4.4# bundler --version
Bundler version 2.4.12
bash-4.4#
bash-4.4# bundle --version
Bundler version 2.4.12
bash-4.4#



bash-4.4# cat Gemfile
source "https://rubygems.org"

source "https://rubygems.org" do
  gem 'kitchen-terraform', '~> 7.0'
end

source "https://packagecloud.io/cinc-project/stable" do
  gem 'cinc-auditor-bin', '5.21.29'
end
bash-4.4#


bash-4.4# bundle install --jobs=8
```

### Which command did you run?

`bash-4.4# bundle install --jobs=8`

### What were you expecting to happen?

Actually, I don't expect those warnings / errors.

### What actually happened?

Warnings printed to console:
```

.......
Fetching gem metadata from https://packagecloud.io/cinc-project/stable/.
Retrying dependency api due to error (2/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.

Retrying dependency api due to error (3/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.

Retrying dependency api due to error (4/4): Bundler::MarshalError TypeError: Unexpected class FalseClass present in marshaled data. Only Symbol, TrueClass, String, Array, Hash, Gem::Version, Gem::Specification are allowed.


```

### PR DESCRIPTION
**Pull request by indirect**
fixes #6614, maybe

### CODE REVIEW

File: bundler/spec/bundler/bundler_spec.rb [hunk 1]
```diff
@@ -28,14 +28,41 @@
...
 
-    it "loads Gem::Version" do
-      gem_version = Gem::Version.new("3.7.2")
-      data = Marshal.dump(gem_version)
-      expect(Bundler.safe_load_marshal(data)).to eq(gem_version)
-    end
-
     it "loads Gem::Specification" do
-      gem_spec = Gem::Specification.new("name", "3.7.2")
+      gem_spec = Gem::Specification.new do |s|
```
Comment (segiddins) [inline bundler/spec/bundler/bundler_spec.rb:32]: nit: this doesn't test `false` ?

**Review by segiddins — APPROVED** [review]

**Review by simi — COMMENTED** [review]
:+1:

**Comment by hsbt** [comment]
👍  @indirect Thanks!

**Comment by stanhu** [comment]
@deivid-rodriguez Do you think you could release RubyGems 3.4.13 to address this? It seems our `bundle install` now fails with 3.4.12 because of this bug.

**Comment by simi** [comment]
@stanhu release is in progress https://github.com/rubygems/rubygems/pull/6666

**Comment by stanhu** [comment]
Thanks!

**Comment by jakerocksfour** [comment]
When you all gonna fix this crap

**Pull request merged**